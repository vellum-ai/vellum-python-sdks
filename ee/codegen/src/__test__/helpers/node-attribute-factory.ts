import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
