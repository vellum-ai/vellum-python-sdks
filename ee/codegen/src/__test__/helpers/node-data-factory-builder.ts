import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
