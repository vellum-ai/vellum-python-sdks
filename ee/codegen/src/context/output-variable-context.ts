import { VellumVariable } from "vellum-ai/api/types";

export declare namespace OutputVariableContext {
  export type Args = {
    outputVariableData: VellumVariable;
  };
}

export class OutputVariableContext {
  private readonly outputVariableData: VellumVariable;
  public readonly name: string;

  constructor({ outputVariableData }: OutputVariableContext.Args) {
    this.outputVariableData = outputVariableData;
    this.name = outputVariableData.key;
  }

  public getOutputVariableId(): string {
    return this.outputVariableData.id;
  }

  public getOutputVariableData(): VellumVariable {
    return this.outputVariableData;
  }

  public getRawName(): string {
    return this.outputVariableData.key;
  }
}
