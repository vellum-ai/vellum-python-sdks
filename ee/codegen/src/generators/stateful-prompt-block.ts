import { python } from "@fern-api/python-ast";
import { ClassInstantiation } from "@fern-api/python-ast/ClassInstantiation";
import { MethodArgument } from "@fern-api/python-ast/MethodArgument";
import { isNil } from "lodash";

import { VELLUM_CLIENT_MODULE_PATH } from "src/constants";
import {
  BasePromptBlock,
  PromptTemplateBlockExcludingFunctionDefinition,
} from "src/generators/base-prompt-block";
import {
  ChatMessagePromptTemplateBlock,
  JinjaPromptTemplateBlock,
  PlainTextPromptTemplateBlock,
  RichTextPromptTemplateBlock,
  VariablePromptTemplateBlock,
} from "src/types/vellum";

// Flesh out unit tests for various prompt configurations
// https://app.shortcut.com/vellum/story/5249
export class StatefulPromptBlock extends BasePromptBlock<PromptTemplateBlockExcludingFunctionDefinition> {
  protected generateAstNode(
    promptBlock: PromptTemplateBlockExcludingFunctionDefinition
  ): ClassInstantiation {
    switch (promptBlock.blockType) {
      case "JINJA":
        return this.generateJinjaPromptBlock(promptBlock);
      case "CHAT_MESSAGE":
        return this.generateChatMessagePromptBlock(promptBlock);
      case "VARIABLE":
        return this.generateVariablePromptBlock(promptBlock);
      case "RICH_TEXT":
        return this.generateRichTextPromptBlock(promptBlock);
      case "PLAIN_TEXT":
        return this.generatePlainTextPromptBlock(promptBlock);
    }
  }

  private getPromptBlockRef(
    promptBlock: PromptTemplateBlockExcludingFunctionDefinition
  ): python.Reference {
    let pathName;
    switch (promptBlock.blockType) {
      case "JINJA":
        pathName = "JinjaPromptBlock";
        break;
      case "CHAT_MESSAGE":
        pathName = "ChatMessagePromptBlock";
        break;
      case "VARIABLE":
        pathName = "VariablePromptBlock";
        break;
      case "RICH_TEXT":
        pathName = "RichTextPromptBlock";
        break;
      case "PLAIN_TEXT":
        pathName = "PlainTextPromptBlock";
        break;
    }
    return python.reference({
      name: pathName,
      modulePath: VELLUM_CLIENT_MODULE_PATH,
    });
  }

  private generateJinjaPromptBlock(
    promptBlock: JinjaPromptTemplateBlock
  ): python.ClassInstantiation {
    const classArgs: MethodArgument[] = [
      ...this.constructCommonClassArguments(promptBlock),
    ];

    if (promptBlock.properties.template) {
      classArgs.push(
        new MethodArgument({
          name: "template",
          value: python.TypeInstantiation.str(promptBlock.properties.template, {
            multiline: true,
            startOnNewLine: true,
            endWithNewLine: true,
          }),
        })
      );
    } else {
      classArgs.push(
        new MethodArgument({
          name: "template",
          value: python.TypeInstantiation.str(""),
        })
      );
    }

    const jinjaBlock = python.instantiateClass({
      classReference: this.getPromptBlockRef(promptBlock),
      arguments_: classArgs,
    });

    this.inheritReferences(jinjaBlock);
    return jinjaBlock;
  }

  private generateChatMessagePromptBlock(
    promptBlock: ChatMessagePromptTemplateBlock
  ): python.ClassInstantiation {
    const classArgs: MethodArgument[] = [
      ...this.constructCommonClassArguments(promptBlock),
    ];

    if (promptBlock.properties.chatRole) {
      classArgs.push(
        new MethodArgument({
          name: "chat_role",
          value: python.TypeInstantiation.str(promptBlock.properties.chatRole),
        })
      );
    }

    if (!isNil(promptBlock.properties.chatSource)) {
      classArgs.push(
        new MethodArgument({
          name: "chat_source",
          value: python.TypeInstantiation.str(
            promptBlock.properties.chatSource
          ),
        })
      );
    }

    if (promptBlock.properties.chatMessageUnterminated) {
      classArgs.push(
        new MethodArgument({
          name: "chat_message_unterminated",
          value: python.TypeInstantiation.bool(
            promptBlock.properties.chatMessageUnterminated
          ),
        })
      );
    }

    const childBlocks = promptBlock.properties.blocks.filter(
      (
        block
      ): block is Exclude<
        PromptTemplateBlockExcludingFunctionDefinition,
        PlainTextPromptTemplateBlock
      > => block.blockType !== "FUNCTION_DEFINITION"
    );
    classArgs.push(
      new MethodArgument({
        name: "blocks",
        value: python.TypeInstantiation.list(
          childBlocks.map((block) => {
            return this.generateAstNode(block);
          })
        ),
      })
    );

    const chatBlock = python.instantiateClass({
      classReference: this.getPromptBlockRef(promptBlock),
      arguments_: classArgs,
    });

    this.inheritReferences(chatBlock);
    return chatBlock;
  }

  private generateVariablePromptBlock(
    promptBlock: VariablePromptTemplateBlock
  ): python.ClassInstantiation {
    const classArgs: MethodArgument[] = [
      ...this.constructCommonClassArguments(promptBlock),
    ];
    const inputVariableName =
      this.inputVariableNameById?.[promptBlock.inputVariableId] ??
      promptBlock.inputVariableId;

    classArgs.push(
      new MethodArgument({
        name: "input_variable",
        value: python.TypeInstantiation.str(inputVariableName),
      })
    );

    const variableBlock = python.instantiateClass({
      classReference: this.getPromptBlockRef(promptBlock),
      arguments_: classArgs,
    });

    this.inheritReferences(variableBlock);
    return variableBlock;
  }

  private generatePlainTextPromptBlock(
    promptBlock: PlainTextPromptTemplateBlock
  ): python.ClassInstantiation {
    const classArgs: MethodArgument[] = [
      ...this.constructCommonClassArguments(promptBlock),
    ];

    const nonEmpty = promptBlock.text !== "";
    classArgs.push(
      new MethodArgument({
        name: "text",
        value: python.TypeInstantiation.str(promptBlock.text, {
          multiline: nonEmpty,
          startOnNewLine: nonEmpty,
          endWithNewLine: nonEmpty,
        }),
      })
    );

    const plainBlock = python.instantiateClass({
      classReference: this.getPromptBlockRef(promptBlock),
      arguments_: classArgs,
    });

    this.inheritReferences(plainBlock);
    return plainBlock;
  }

  private generateRichTextPromptBlock(
    promptBlock: RichTextPromptTemplateBlock
  ): python.ClassInstantiation {
    const classArgs: MethodArgument[] = [
      ...this.constructCommonClassArguments(promptBlock),
    ];

    classArgs.push(
      new MethodArgument({
        name: "blocks",
        value: python.TypeInstantiation.list(
          promptBlock.blocks.map((block) => {
            if (block.blockType === "VARIABLE") {
              return this.generateVariablePromptBlock(block);
            } else {
              return this.generatePlainTextPromptBlock(block);
            }
          })
        ),
      })
    );

    const richBlock = python.instantiateClass({
      classReference: this.getPromptBlockRef(promptBlock),
      arguments_: classArgs,
    });

    this.inheritReferences(richBlock);
    return richBlock;
  }
}
