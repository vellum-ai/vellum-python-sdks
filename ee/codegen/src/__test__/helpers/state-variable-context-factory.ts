import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
