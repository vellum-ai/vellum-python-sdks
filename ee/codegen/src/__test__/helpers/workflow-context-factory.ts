import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
