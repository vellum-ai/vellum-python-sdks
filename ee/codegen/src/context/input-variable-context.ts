import { isEmpty, isNil } from "lodash";
import { CodeResourceDefinition, VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context/workflow-context";
import {
  removeEscapeCharacters,
  toValidPythonIdentifier,
} from "src/utils/casing";
import { getGeneratedInputsModulePath } from "src/utils/paths";

export declare namespace InputVariableContext {
  export type Args = {
    inputVariableData: VellumVariable;
    workflowContext: WorkflowContext;
  };
}

export class InputVariableContext {
  private readonly workflowContext: WorkflowContext;
  private readonly inputVariableData: VellumVariable;
  public readonly definition: CodeResourceDefinition;

  public readonly name: string;

  constructor({
    inputVariableData,
    workflowContext,
  }: InputVariableContext.Args) {
    this.workflowContext = workflowContext;
    this.inputVariableData = inputVariableData;
    this.definition = {
      name: "Inputs",
      module: getGeneratedInputsModulePath(workflowContext),
    };

    this.name = this.generateSanitizedInputVariableName();
  }

  public getInputVariableId(): string {
    return this.inputVariableData.id;
  }

  public getInputVariableData(): VellumVariable {
    return this.inputVariableData;
  }

  public getRawName(): string {
    // This is for an edge case where there are escape characters in the variable
    return removeEscapeCharacters(this.inputVariableData.key);
  }

  private generateSanitizedInputVariableName(): string {
    const defaultName = "input_";
    const rawInputVariableName = this.inputVariableData.key;

    const initialInputVariableName =
      !isNil(rawInputVariableName) && !isEmpty(rawInputVariableName)
        ? toValidPythonIdentifier(rawInputVariableName, "input")
        : defaultName;

    // Deduplicate the input variable name if it's already in use
    let sanitizedName = initialInputVariableName;
    let numRenameAttempts = 0;
    while (this.workflowContext.isInputVariableNameUsed(sanitizedName)) {
      sanitizedName = `${initialInputVariableName}${
        initialInputVariableName.endsWith("_") ? "" : "_"
      }${numRenameAttempts + 1}`;
      numRenameAttempts += 1;
    }

    return sanitizedName;
  }
}
