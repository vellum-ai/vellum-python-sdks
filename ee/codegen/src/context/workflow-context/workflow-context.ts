import { MlModels } from "vellum-ai/api/resources/mlModels/client/Client";

import { GENERATED_WORKFLOW_MODULE_NAME } from "src/constants";
import { InputVariableContext } from "src/context/input-variable-context";
import { BaseNodeContext } from "src/context/node-context/base";
import { OutputVariableContext } from "src/context/output-variable-context";
import { PortContext } from "src/context/port-context";
import { generateSdkModulePaths } from "src/context/workflow-context/sdk-module-paths";
import { SDK_MODULE_PATHS } from "src/context/workflow-context/types";
import { WorkflowOutputContext } from "src/context/workflow-output-context";
import {
  BaseCodegenError,
  CodegenErrorSeverity,
  NodeDefinitionGenerationError,
  NodeNotFoundError,
  NodePortGenerationError,
  NodePortNotFoundError,
  WorkflowGenerationError,
  WorkflowInputGenerationError,
  WorkflowOutputGenerationError,
} from "src/generators/errors";
import { BaseNode } from "src/generators/nodes/bases";
import {
  EntrypointNode,
  WorkflowDataNode,
  WorkflowEdge,
  WorkflowRawData,
} from "src/types/vellum";
import { createPythonClassName } from "src/utils/casing";

type InputVariableContextsById = Map<string, InputVariableContext>;

type OutputVariableContextsById = Map<string, OutputVariableContext>;

type NodeContextsByNodeId = Map<string, BaseNodeContext<WorkflowDataNode>>;

// A mapping between source handle ids and port contexts
type PortContextById = Map<string, PortContext>;

export declare namespace WorkflowContext {
  export type Args = {
    absolutePathToOutputDirectory: string;
    moduleName: string;
    workflowClassName: string;
    globalInputVariableContextsById?: InputVariableContextsById;
    globalNodeContextsByNodeId?: NodeContextsByNodeId;
    globalOutputVariableContextsById?: OutputVariableContextsById;
    parentNode?: BaseNode<WorkflowDataNode, BaseNodeContext<WorkflowDataNode>>;
    workflowsSdkModulePath?: readonly string[];
    portContextByName?: PortContextById;
    vellumApiKey: string;
    workflowRawData: WorkflowRawData;
    strict: boolean;
    codeExecutionNodeCodeRepresentationOverride: "STANDALONE" | "INLINE";
    disableFormatting: boolean;
    classNames?: Set<string>;
  };
}

export class WorkflowContext {
  public readonly absolutePathToOutputDirectory: string;
  public readonly modulePath: string[];
  public readonly moduleName: string;
  public readonly label: string | undefined;
  public readonly workflowClassName: string;

  // Maps workflow input variable IDs to the input variable
  // Tracks local and global contexts in the case of nested workflows.
  public readonly inputVariableContextsById: InputVariableContextsById;
  public readonly globalInputVariableContextsById: InputVariableContextsById;

  // Maps workflow output variable IDs to the output variable
  // Tracks local and global contexts in the case of nested workflows.
  public readonly outputVariableContextsById: OutputVariableContextsById;
  public readonly globalOutputVariableContextsById: OutputVariableContextsById;

  public readonly strict: boolean;

  // Track what input variables names are used within this workflow so that we can ensure name uniqueness when adding
  // new input variables.
  private readonly inputVariableNames: Set<string> = new Set();

  // Maps node IDs to a mapping of output IDs to output names.
  // Tracks local and global contexts in the case of nested workflows.
  public readonly nodeContextsByNodeId: NodeContextsByNodeId;
  public readonly globalNodeContextsByNodeId: NodeContextsByNodeId;

  // Track what node module names are used within this workflow so that we can ensure name uniqueness when adding
  // new nodes.
  private readonly nodeModuleNames: Set<string> = new Set();

  // A list of all outputs this workflow produces
  public readonly workflowOutputContexts: WorkflowOutputContext[] = [];

  // Track what output variables names are used within this workflow so that we can ensure name uniqueness when adding
  // new output variables.
  private readonly outputVariableNames: Set<string> = new Set();

  // If this workflow is a nested workflow belonging to a node, track that node's context here.
  public readonly parentNode?: BaseNode<
    WorkflowDataNode,
    BaseNodeContext<WorkflowDataNode>
  >;

  // The entrypoint node for this workflow
  private entrypointNode: EntrypointNode | undefined;

  public readonly sdkModulePathNames: SDK_MODULE_PATHS;

  public readonly portContextById: PortContextById;

  // Used by the vellum api client
  public readonly vellumApiKey: string;
  private readonly mlModelNamesById: Record<string, string> = {};
  private readonly errors: BaseCodegenError[] = [];

  public readonly workflowRawData: WorkflowRawData;

  public readonly codeExecutionNodeCodeRepresentationOverride:
    | "STANDALONE"
    | "INLINE";

  public readonly disableFormatting: boolean;

  // Track what class names are used within this workflow so that we can ensure name uniqueness
  public readonly classNames: Set<string>;

  constructor({
    absolutePathToOutputDirectory,
    moduleName,
    workflowClassName,
    globalInputVariableContextsById,
    globalNodeContextsByNodeId,
    globalOutputVariableContextsById,
    parentNode,
    workflowsSdkModulePath = ["vellum", "workflows"] as const,
    portContextByName,
    vellumApiKey,
    workflowRawData,
    strict,
    codeExecutionNodeCodeRepresentationOverride,
    disableFormatting,
    classNames,
  }: WorkflowContext.Args) {
    this.absolutePathToOutputDirectory = absolutePathToOutputDirectory;
    this.moduleName = moduleName;
    this.modulePath = parentNode
      ? [
          ...parentNode.nodeContext.nodeModulePath,
          GENERATED_WORKFLOW_MODULE_NAME,
        ]
      : [this.moduleName, GENERATED_WORKFLOW_MODULE_NAME];
    this.workflowClassName = workflowClassName;
    this.vellumApiKey = vellumApiKey;

    this.inputVariableContextsById = new Map();
    this.globalInputVariableContextsById =
      globalInputVariableContextsById ?? new Map();

    this.nodeContextsByNodeId = new Map();
    this.globalNodeContextsByNodeId = globalNodeContextsByNodeId ?? new Map();

    this.portContextById = portContextByName ?? new Map();

    this.parentNode = parentNode;

    this.sdkModulePathNames = generateSdkModulePaths(workflowsSdkModulePath);
    this.workflowRawData = workflowRawData;

    this.strict = strict;
    this.errors = [];

    this.codeExecutionNodeCodeRepresentationOverride =
      codeExecutionNodeCodeRepresentationOverride;

    this.disableFormatting = disableFormatting;

    this.outputVariableContextsById = new Map();
    this.globalOutputVariableContextsById =
      globalOutputVariableContextsById ?? new Map();

    this.classNames = classNames ?? new Set<string>();
  }

  /* Create a new workflow context for a nested workflow from its parent */
  public createNestedWorkflowContext({
    parentNode,
    workflowClassName,
    workflowRawData,
    classNames,
  }: {
    parentNode: BaseNode<WorkflowDataNode, BaseNodeContext<WorkflowDataNode>>;
    workflowClassName: string;
    workflowRawData: WorkflowRawData;
    classNames?: Set<string>;
  }) {
    return new WorkflowContext({
      absolutePathToOutputDirectory: this.absolutePathToOutputDirectory,
      moduleName: this.moduleName,
      workflowClassName: workflowClassName,
      globalInputVariableContextsById: this.globalInputVariableContextsById,
      globalNodeContextsByNodeId: this.globalNodeContextsByNodeId,
      globalOutputVariableContextsById: this.globalOutputVariableContextsById,
      parentNode,
      workflowsSdkModulePath: this.sdkModulePathNames.WORKFLOWS_MODULE_PATH,
      vellumApiKey: this.vellumApiKey,
      workflowRawData,
      codeExecutionNodeCodeRepresentationOverride:
        this.codeExecutionNodeCodeRepresentationOverride,
      strict: this.strict,
      disableFormatting: this.disableFormatting,
      classNames,
    });
  }

  public getEntrypointNode(): EntrypointNode {
    if (this.entrypointNode) {
      return this.entrypointNode;
    }

    const entrypointNodes = this.workflowRawData.nodes.filter(
      (n): n is EntrypointNode => n.type === "ENTRYPOINT"
    );
    if (entrypointNodes.length > 1) {
      throw new WorkflowGenerationError("Multiple entrypoint nodes found");
    }

    const entrypointNode = entrypointNodes[0];
    if (!entrypointNode) {
      throw new WorkflowGenerationError("Entrypoint node not found");
    }
    this.entrypointNode = entrypointNode;

    return this.entrypointNode;
  }

  public getEntrypointNodeEdges(): WorkflowEdge[] {
    const entrypointNodeId = this.getEntrypointNode().id;
    return this.workflowRawData.edges.filter(
      (edge) => edge.sourceNodeId === entrypointNodeId
    );
  }

  public getEdgesByPortId(): Map<string, WorkflowEdge[]> {
    const edgesByPortId = new Map<string, WorkflowEdge[]>();
    this.workflowRawData.edges.forEach((edge) => {
      const portId = edge.sourceHandleId;
      const edges = edgesByPortId.get(portId) ?? [];
      edges.push(edge);
      edgesByPortId.set(portId, edges);
    });
    return edgesByPortId;
  }

  public isInputVariableNameUsed(inputVariableName: string): boolean {
    return this.inputVariableNames.has(inputVariableName);
  }

  private addUsedInputVariableName(inputVariableName: string): void {
    this.inputVariableNames.add(inputVariableName);
  }

  public addInputVariableContext(
    inputVariableContext: InputVariableContext
  ): void {
    const inputVariableId = inputVariableContext.getInputVariableId();

    if (this.globalInputVariableContextsById.get(inputVariableId)) {
      throw new WorkflowInputGenerationError(
        `Input variable context already exists for input variable ID: ${inputVariableId}`
      );
    }

    this.inputVariableContextsById.set(inputVariableId, inputVariableContext);
    this.globalInputVariableContextsById.set(
      inputVariableId,
      inputVariableContext
    );
    this.addUsedInputVariableName(inputVariableContext.name);
  }

  public findInputVariableContextById(
    inputVariableId: string
  ): InputVariableContext | undefined {
    return this.globalInputVariableContextsById.get(inputVariableId);
  }

  public findOutputVariableContextById(
    outputVariableId: string
  ): OutputVariableContext | undefined {
    return this.globalOutputVariableContextsById.get(outputVariableId);
  }

  public addOutputVariableContext(
    outputVariableContext: OutputVariableContext
  ): void {
    const outputVariableId = outputVariableContext.getOutputVariableId();

    if (this.globalOutputVariableContextsById.get(outputVariableId)) {
      throw new WorkflowOutputGenerationError(
        `Output variable context already exists for output variable ID: ${outputVariableId}`
      );
    }

    this.outputVariableContextsById.set(
      outputVariableId,
      outputVariableContext
    );
    this.globalOutputVariableContextsById.set(
      outputVariableId,
      outputVariableContext
    );

    // This was added from the workflow output context. We should remove this once terminal node data is removed from data
    this.addUsedOutputVariableName(outputVariableContext.name);
  }

  public getOutputVariableContextById(
    outputVariableId: string
  ): OutputVariableContext {
    const outputVariableContext =
      this.findOutputVariableContextById(outputVariableId);

    if (!outputVariableContext) {
      throw new WorkflowInputGenerationError(
        `Output variable context not found for ID: ${outputVariableId}`
      );
    }

    return outputVariableContext;
  }

  public findInputVariableContextByRawName(
    rawName: string
  ): InputVariableContext | undefined {
    return Array.from(this.inputVariableContextsById.values()).find(
      (inputContext) => inputContext.getRawName() === rawName
    );
  }

  public isOutputVariableNameUsed(outputVariableName: string): boolean {
    return this.outputVariableNames.has(outputVariableName);
  }

  private addUsedOutputVariableName(outputVariableName: string): void {
    this.outputVariableNames.add(outputVariableName);
  }

  public addWorkflowOutputContext(
    workflowOutputContext: WorkflowOutputContext
  ): void {
    this.workflowOutputContexts.push(workflowOutputContext);
  }

  public isNodeModuleNameUsed(nodeModuleName: string): boolean {
    return this.nodeModuleNames.has(nodeModuleName);
  }

  private addUsedNodeModuleName(nodeModuleName: string): void {
    this.nodeModuleNames.add(nodeModuleName);
  }

  public addNodeContext(nodeContext: BaseNodeContext<WorkflowDataNode>): void {
    const nodeId = nodeContext.nodeData.id;

    if (this.globalNodeContextsByNodeId.get(nodeId)) {
      throw new NodeDefinitionGenerationError(
        `Node context already exists for node ID: ${nodeId}`
      );
    }

    this.nodeContextsByNodeId.set(nodeId, nodeContext);
    this.globalNodeContextsByNodeId.set(nodeId, nodeContext);
    this.addUsedNodeModuleName(nodeContext.nodeModuleName);
  }

  public findNodeContext(
    nodeId: string
  ): BaseNodeContext<WorkflowDataNode> | undefined {
    const nodeContext = this.globalNodeContextsByNodeId.get(nodeId);

    if (!nodeContext) {
      this.addError(
        new NodeNotFoundError(
          `Failed to find node with id '${nodeId}'`,
          "WARNING"
        )
      );
    }

    return nodeContext;
  }

  public addPortContext(portContext: PortContext): void {
    const portId = portContext.portId;

    if (this.portContextById.get(portId)) {
      throw new NodePortGenerationError(
        `Port context already exists for port id: ${portId}`
      );
    }
    this.portContextById.set(portId, portContext);
  }

  public getPortContextById(portId: string): PortContext {
    const portContext: PortContext | undefined =
      this.portContextById.get(portId);

    if (!portContext) {
      throw new NodePortNotFoundError(
        `Port context not found for port id: ${portId}`
      );
    }

    return portContext;
  }

  public async getMLModelNameById(mlModelId: string): Promise<string> {
    const mlModelName = this.mlModelNamesById[mlModelId];
    if (mlModelName) {
      return mlModelName;
    }

    const mlModel = await new MlModels({ apiKey: this.vellumApiKey }).retrieve(
      mlModelId
    );

    this.mlModelNamesById[mlModelId] = mlModel.name;
    return mlModel.name;
  }

  public addWorkflowEdges(edges: WorkflowEdge[]): void {
    this.workflowRawData.edges.push(...edges);
  }

  public addError(error: BaseCodegenError): void {
    if (this.strict) {
      throw error;
    } else {
      error.log();
    }

    const errorExists = this.errors.some(
      (existingError) => existingError.message === error.message
    );

    if (!errorExists) {
      this.errors.push(error);
    }
  }

  public getErrors(severity?: CodegenErrorSeverity): BaseCodegenError[] {
    const allErrors = [...this.errors];
    if (!severity) {
      return allErrors;
    }
    return allErrors.filter((error) => error.severity === severity);
  }

  public isClassNameUsed(className: string): boolean {
    return this.classNames.has(className);
  }

  private addUsedClassName(className: string): void {
    this.classNames.add(className);
  }

  public getUniqueClassName(baseName: string): string {
    let sanitizedName = createPythonClassName(baseName);
    let numRenameAttempts = 0;

    if (!this.isClassNameUsed(sanitizedName)) {
      this.addUsedClassName(sanitizedName);
      return sanitizedName;
    }

    while (this.isClassNameUsed(sanitizedName)) {
      numRenameAttempts += 1;
      sanitizedName = `${createPythonClassName(baseName)}${numRenameAttempts}`;
    }

    this.addUsedClassName(sanitizedName);
    return sanitizedName;
  }
}
