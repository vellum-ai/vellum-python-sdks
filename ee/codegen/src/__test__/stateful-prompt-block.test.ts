import { Writer } from "@fern-api/python-ast/core/Writer";

import { workflowContextFactory } from "./helpers";

import { WorkflowContext } from "src/context/workflow-context";
import { StatefulPromptBlock } from "src/generators/stateful-prompt-block";

describe("StatefulPromptBlock", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;

  beforeEach(() => {
    workflowContext = workflowContextFactory();
    writer = new Writer();
  });

  describe("JINJA", () => {
    it("should generate a basic jinja block", async () => {
      const block = new StatefulPromptBlock({
        workflowContext,
        promptBlock: {
          id: "1",
          blockType: "JINJA",
          state: "ENABLED",
          properties: {
            template: "Hello, {{ name }}!",
          },
        },
        inputVariableNameById: {},
      });

      block.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should generate a jinja block with a cache config", async () => {
      const block = new StatefulPromptBlock({
        workflowContext,
        promptBlock: {
          id: "1",
          blockType: "JINJA",
          state: "ENABLED",
          properties: {
            template: "Hello, {{ name }}!",
          },
          cacheConfig: {
            type: "EPHEMERAL",
          },
        },
        inputVariableNameById: {},
      });

      block.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
    it("should handle double quotes in jinja template", async () => {
      const block = new StatefulPromptBlock({
        workflowContext,
        promptBlock: {
          id: "1",
          blockType: "JINJA",
          state: "ENABLED",
          properties: {
            template: '"Hello" "World"',
          },
        },
        inputVariableNameById: {},
      });

      block.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("PLAIN_TEXT", () => {
    it("should generate regular quotes for empty text", async () => {
      const block = new StatefulPromptBlock({
        workflowContext,
        promptBlock: {
          id: "1",
          blockType: "PLAIN_TEXT",
          state: "ENABLED",
          text: "",
        },
        inputVariableNameById: {},
      });

      block.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });
});
