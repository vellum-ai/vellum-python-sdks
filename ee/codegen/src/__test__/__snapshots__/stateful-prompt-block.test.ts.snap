// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatefulPromptBlock > JINJA > should generate a basic jinja block 1`] = `
"JinjaPromptBlock(template="""Hello, {{ name }}!""")
"
`;

exports[`StatefulPromptBlock > JINJA > should generate a jinja block with a cache config 1`] = `
"JinjaPromptBlock(
    cache_config=EphemeralPromptCacheConfig(type="EPHEMERAL"),
    template="""Hello, {{ name }}!""",
)
"
`;

exports[`StatefulPromptBlock > JINJA > should handle double quotes in jinja template 1`] = `
"JinjaPromptBlock(template="""\\"Hello\\" \\"World\\"""")
"
`;

exports[`StatefulPromptBlock > PLAIN_TEXT > should generate regular quotes for empty text 1`] = `
"PlainTextPromptBlock(text="")
"
`;
